homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,138 @@
-    vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+    vol.Required(CONF_HOST): cv.string,
     vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
     vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+    vol.Optional(CONF_ADBKEY, default=DEFAULT_ADBKEY): cv.string,
+    vol.Optional(CONF_GET_SOURCE, default=DEFAULT_GET_SOURCE): cv.boolean,
+    vol.Optional(CONF_GET_SOURCES, default=DEFAULT_GET_SOURCES): cv.boolean
 })
 
+PACKAGE_LAUNCHER = "com.amazon.tv.launcher"
+PACKAGE_SETTINGS = "com.amazon.tv.settings"
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+
+# pylint: disable=protected-access
+def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the FireTV platform."""
+    host = '{0}:{1}'.format(config.get(CONF_HOST), config.get(CONF_PORT))
     name = config.get(CONF_NAME)
-    ssl = config.get(CONF_SSL)
-    proto = 'https' if ssl else 'http'
-    host = config.get(CONF_HOST)
-    port = config.get(CONF_PORT)
-    device_id = config.get(CONF_DEVICE)
-
-    try:
-        response = requests.get(
-            DEVICE_LIST_URL.format(proto, host, port)).json()
-        if device_id in response[CONF_DEVICES].keys():
-            add_entities([FireTVDevice(proto, host, port, device_id, name)])
-            _LOGGER.info("Device %s accessible and ready for control",
-                         device_id)
-        else:
-            _LOGGER.warning("Device %s is not registered with firetv-server",
-                            device_id)
-    except requests.exceptions.RequestException:
-        _LOGGER.error("Could not connect to firetv-server at %s", host)
-
-
-class FireTV:
-    """The firetv-server client.
-
-    Should a native Python 3 ADB module become available, python-firetv can
-    support Python 3, it can be added as a dependency, and this class can be
-    dispensed of.
-
-    For now, it acts as a client to the firetv-server HTTP server (which must
-    be running via Python 2).
-    """
-
-    def __init__(self, proto, host, port, device_id):
-        """Initialize the FireTV server."""
-        self.proto = proto
-        self.host = host
-        self.port = port
-        self.device_id = device_id
-
-    @property
-    def state(self):
-        """Get the device state. An exception means UNKNOWN state."""
+    adbkey = config.get(CONF_ADBKEY)
+    get_source = config.get(CONF_GET_SOURCE)
+    get_sources = config.get(CONF_GET_SOURCES)
+
+    device = FireTVDevice(host, name, adbkey, get_source, get_sources)
+    adb_log = " using adbkey='{0}'".format(adbkey) if adbkey else ""
+    if not device._firetv._adb:
+        _LOGGER.warning("Could not connect to Fire TV at %s%s", host, adb_log)
+
+        # Configuration troubleshooting for `adbkey`
+        if adbkey:
+            # Check whether the key files exist
+            if not os.path.exists(adbkey):
```

Comment:
Is something like this what you have in mind?  

https://github.com/home-assistant/home-assistant/blob/69c04697d24470c90535410860472efc6cac6f36/homeassistant/components/media_player/vizio.py#L56-L67

https://github.com/home-assistant/home-assistant/blob/69c04697d24470c90535410860472efc6cac6f36/homeassistant/components/media_player/vizio.py#L182-L184

Reply:
No, I mean that we should check if the files exist in the platform config schema. Also remove the default adb key (empty string) from the config schema. The `FireTV` class already has a default for it, so we shouldn't pass anything if it's not specified in the config schema.
```py
def has_adb_files(value):
    """Check that adb files exist."""
    priv_key = value
    pub_key = '{}.pub'.format(value)
    cv.isfile(pub_key)
    return cv.isfile(priv_key)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    ...
    vol.Optional(CONF_ADBKEY): has_adb_files,
    ...
})
```

But it would also be good to check connection status before finishing platform set up. If we think that the connection can succeed later, we can raise PlatformNotReady error. If we don't think that, we should just log an error and return.

Reply:
Fixed in https://github.com/home-assistant/home-assistant/pull/17767/commits/80daea67be3aa84a3e511aff54b95d7b3e4e18e7